R/susie_ss.R

```diff
@@ -0,0 +1,106 @@
+#' sum_l b_l is a p vector of effects to be estimated.
+#' The assumption is that each b_l has exactly one non-zero element, with all elements
+#' equally likely to be non-zero. The prior on the non-zero element is N(0,var=sa2*s2).
+#' Only the summary statistics t(X)Y and t(X)X are available.
+#' @param XtX a p by p matrix, t(X)X
+#' @param XtY a p vector
+#' @param L maximum number of non-zero effects
+#' @param prior_variance the scaled prior variance (vector of length L, or scalar. In latter case gets repeated L times )
+#' @param residual_variance the residual variance (defaults to variance of Y)
+#' @param estimate_prior_variance indicates whether to estimate prior (currently not recommended as not working as well)
+#' @param max_iter maximum number of iterations to perform
+#' @param s_init a previous susie fit with which to initialize
+#' @return a susie fit, which is a list with some or all of the following elements\cr
+#' \item{alpha}{an L by p matrix of posterior inclusion probabilites}
+#' \item{mu}{an L by p matrix of posterior means (conditional on inclusion)}
+#' \item{mu2}{an L by p matrix of posterior second moments (conditional on inclusion)}
+#' \item{XtXr}{an p vector of t(X) times fitted values, the fitted values equal to X times colSums(alpha*mu))}
+#' \item{sigma2}{residual variance}
+#' \item{sa2}{scaled prior variance; ie prior variance is sigma2*sa2}
+#' @examples
+#' set.seed(1)
+#' n = 1000
+#' p = 1000
+#' beta = rep(0,p)
+#' beta[1] = 1
+#' beta[2] = 1
+#' beta[3] = 1
+#' beta[4] = 1
+#' X = matrix(rnorm(n*p),nrow=n,ncol=p)
+#' y = X %*% beta + rnorm(n)
+#' res =susie_ss(XtX,Xty,L=10)
+#' coef(res)
+#' plot(y,predict(res))
+#' @export
+susie_ss = function(XtX,Xty,L=10,prior_variance=0.2,residual_variance=NULL,estimate_prior_variance = FALSE, max_iter=100,s_init = NULL, verbose=FALSE){
+  # Check input XtX.
+  if (!is.double(XtX) || !is.matrix(XtX))
+    stop("Input XtX must be a double-precision matrix")
+  p = ncol(XtX)
+
+  # initialize susie fit
+  if(!is.null(s_init)){
+    if(!missing(L) || !missing(prior_variance) || !missing(residual_variance))
+      stop("if provide s_init then L, sa2 and sigma2 must not be provided")
+    s = s_init
+  } else {
+
+    if(is.null(residual_variance)){
+      residual_variance = readline(prompt="What is the variance of Y?")
+    }
+    residual_variance= as.numeric(residual_variance) #avoid problems with dimension if entered as matrix
+
+
+    if(length(prior_variance)==1){
+      prior_variance = rep(prior_variance,L)
+    }
+
+    # Check inputs sigma and sa.
+    if (length(residual_variance) != 1)
+      stop("Inputs residual_variance must be scalar")
+    # Check inputs sigma and sa.
+    if (length(prior_variance) != L)
+      stop("Inputs prior_variance must be of length 1 or L")
+
+    #initialize susie fit
+    s = list(alpha=matrix(1/p,nrow=L,ncol=p), mu=matrix(0,nrow=L,ncol=p),
```

Comment:
i think best to include s$intercept = 0 so this way the coef(s) will be the same length for both ss and non-ss versions.